code/modules/urist/modules/GOAI/integrations/base_commander/mob_commander/obstacle_handlers/airlocks.dm

```diff
@@ -0,0 +1,287 @@
+			200,
+			1,
+			FALSE,
+			action_args
+		)
+
+		hack_handled = TRUE
+
+	else if(panel_open && !no_power)	//AttackHand/Bump doesn't not work if the panel is open. Close it first if we're not hacking the airlock and it has power
+		action_key= "[NEED_OBJ_SCREW(A)] for [waypoint]"
+
+		var/list/extra_preconds = base_preconds.Copy()
+
+		open_door_preconds[action_key] = TRUE
+		src.SetState(action_key, FALSE)	//Workaround
+
+		extra_preconds[action_key] = -TRUE
+		extra_preconds[STATE_HASSCREWDRIVER] = TRUE
+
+		var/list/effects = list()
+
+		effects[action_key] = TRUE
+
+		var/list/action_args = list()
+		action_args["target"] = A
+
+		AddAction(
+			action_key,
+			extra_preconds,
+			effects,
+			/datum/goai/mob_commander/proc/HandleScrew,
+			5,
+			1,
+			FALSE,
+			action_args
+		)
+
+	//---- Opening methods: Pick one ----//
+
+	if(no_power || hack_handled)
+		action_key = "[NEED_OBJ_PRY(A)] for [waypoint]"
+
+		open_door_preconds[action_key] = -TRUE
+		open_door_preconds[STATE_HASCROWBAR] = TRUE
+		src.SetState(action_key, FALSE)	//Workaround
+
+		var/list/effects = list()
+		effects[action_key] = TRUE
+
+		var/list/action_args = list()
+		action_args["target"] = A
+
+		AddAction(
+			action_key,
+			open_door_preconds,
+			effects,
+			/datum/goai/mob_commander/proc/HandlePry,
+			15 + rand() - (pawn ? get_dist(A, pawn) : 0),
+			1,
+			FALSE,
+			action_args
+		)
+	//Prying unpowered airlocks is an /alternative/ way of opening them, not a prerequisite step. We need an else here
+	else
+		action_key = "[NEED_OBSTACLE_OPEN(A)] for [waypoint]"
+
+		open_door_preconds[action_key] = -TRUE
+		src.SetState(action_key, FALSE)	//Workaround
+
+		var/list/open_door_effects = list()
+		open_door_effects[action_key] = TRUE
+
+		var/list/action_args = list()
+		action_args["obstruction"] = A
+
+		AddAction(
+			action_key,
+			open_door_preconds,
+			open_door_effects,
+			/datum/goai/mob_commander/proc/HandleAirlockOpen,
+			10 + rand() - (pawn ? get_dist(A, pawn) : 0),
+			1,
+			FALSE,
+			action_args
+		)
+
+	return action_key
+
+
+/datum/goai/mob_commander/proc/HandleAirlockOpen(var/datum/ActionTracker/tracker, var/obj/machinery/door/airlock/obstruction)
+	var/mob/pawn = src.GetPawn()
+	if (!pawn || !istype(pawn))
+		return
+
+	if (!tracker)
+		return
+
+	var/walk_dist = (tracker?.BBSetDefault("StartDist", (ManhattanDistance(get_turf(pawn), obstruction) || 0)) || 0)
+
+	if(tracker.IsOlderThan(src.ai_tick_delay * (10 + walk_dist)))
+		if(src.brain.GetMemoryValue(MEM_OBSTRUCTION, null) == obstruction)
+			src.brain.DropMemory(MEM_OBSTRUCTION)
+		tracker.SetFailed()
+		return
+
+	//If the mob can see the airlock, check if its bolt lights are on
+	if(obstruction in src.brain.perceptions[SENSE_SIGHT_CURR])
+		if(obstruction.locked && obstruction.lights)
+			tracker.SetFailed()
+			src.brain.SetMemory(MEM_OBJ_LOCKED(obstruction), TRUE, 5 MINUTES)
+			if(src.brain.GetMemoryValue(MEM_OBSTRUCTION, null) == obstruction)
```

Comment:
Not specifically directed at you, but this made me realize - we'll probably need to refactor this to store a weakref and resolve it, or the GC will get mad at us.